app/src/main/java/org/simple/clinic/patient/filter/WeightedLevenshteinSearch.kt

```diff
@@ -0,0 +1,72 @@
+package org.simple.clinic.patient.filter
+
+import info.debatty.java.stringsimilarity.CharacterInsDelInterface
+import info.debatty.java.stringsimilarity.CharacterSubstitutionInterface
+import info.debatty.java.stringsimilarity.WeightedLevenshtein
+import io.reactivex.Single
+import org.simple.clinic.patient.PatientSearchResult
+import java.util.UUID
+
+class WeightedLevenshteinSearch(
+    private val minimumSearchTermLength: Int,
+    private val fuzzyStringDistanceCutoff: Float,
+    private val resultsComparator: Comparator<SearchContext>,
+    characterSubstitutionCost: Float,
+    characterInsertionCost: Float,
+    characterDeletionCost: Float
+) : SearchPatientByName {
+
+  init {
+    if (minimumSearchTermLength < 1) {
+      throw AssertionError("Minimum search term length cannot be less than 1")
+    }
+  }
+
+  private val whiteSpaceRegex = Regex("[\\s]")
+
+  private val levenshtein = WeightedLevenshtein(
+      CharacterSubstitutionInterface { _, _ -> characterSubstitutionCost.toDouble() },
+      object : CharacterInsDelInterface {
+        override fun deletionCost(c: Char) = characterDeletionCost.toDouble()
+
+        override fun insertionCost(c: Char) = characterInsertionCost.toDouble()
+      }
+  )
+
+  override fun search(searchTerm: String, names: List<PatientSearchResult.PatientNameAndId>): Single<List<UUID>> {
+    val searchTermParts = nameToSearchableParts(searchTerm)
+        .filter { it.length >= minimumSearchTermLength }
+
+    return if (searchTermParts.isEmpty()) {
+      Single.just(emptyList())
+    } else {
+      val searchContext = names
+          .filter { it.fullName.isNotBlank() }
+          .map {
+            SearchContext(
+                patient = it,
+                nameParts = nameToSearchableParts(it.fullName),
+                searchParts = searchTermParts)
```

Comment:
I'm getting slightly confused with the naming here. The value of `nameToSearchableParts()` is stored in `nameParts`, but there's another property called `searchParts`.

Reply:
Ah, I should probably name this function differently, probably call it `stringToSearchableParts` or something.

`nameParts` is the split parts of the patient name, and `searchParts` is the split parts of the search term.

Reply:
Renamed to `stringToSearchableParts `.

Reply:
Will `searchParts` -> `searchQueryParts` be more understandable? Feel free to do this in your next PR.

Reply:
Actually, yes. Will make the change in my coming PR.